### PR TITLE
Add asynchronous_complete_cumsum operator for XPU

### PR DESCRIPTION
**Pull request by aagalleg**
Depends on: #72 

This PR introduces the `asynchronous_complete_cumsum` operator to fbgemm-xpu, enabling complete cumulative sum computation with a leading zero on Intel XPU devices.

#### Core Implementation

- **XPU Implementation** (`asynchronous_complete_cumsum.cpp/h`): SYCL port of FBGEMM's asynchronous complete cumsum operator that computes cumulative sum with prepended zero (e.g., `[a, b, c]` → `[0, a, a+b, a+b+c]`)
- **Operator Registration** (`ops_registry.cpp`): Registers the operator with PyTorch's dispatch system using conditional schema registration to avoid conflicts
- **Python API** (`ops.py`): Clean Python wrapper function with type hints and documentation for easy integration

#### Infrastructure

- **Build System** (`CMakeLists.txt`): Added implementation to build configuration
- **Tensor Support**: Handles both 1D and 2D tensors (cumsum along last dimension for 2D)
- **Data Types**: Supports int32 and int64 tensors

#### Testing

- **Comprehensive Test Suite** (`test_asynchronous_complete_cumsum.py`):
  - Correctness validation for int32/int64 data types
  - Edge cases (empty tensors)
  - Random input validation with numpy reference
  - XPU device testing with proper error handling

cc: @flezaalv, @manuelhsantana

### CODE REVIEW

File: packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace fbgemm_xpu::utils::torch {
+
+inline bool schemaExists(const std::string& qualified_name) {
+  return c10::Dispatcher::singleton()
+      .findSchema({qualified_name, ""})
+      .has_value();
+}
+
+} // namespace fbgemm_xpu::utils::torch
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h]: add new line in end of file

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    const int64_t global_range = item.get_global_range(0);
+    
+    // Grid-stride loop pattern for scalability
+    for (int64_t i = global_id; i < numel_; i += global_range) {
+        const int32_t target_idx = permute_[i];
+        inversed_permute_[target_idx] = static_cast<int32_t>(i);
+    }
+}
+
+void InvertPermuteKernelInt64::operator()(const sycl::nd_item<1>& item) const {
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/invert_permute_kernel.cpp]: This seems templetable kernel over the operand type. Why don't we do that especially considering that the reference CUDA kernel is doing exactly that [1]?

[1] https://github.com/pytorch/FBGEMM/blob/8744183e4bd7390a674a0d8736bf4d7ef2d0516e/fbgemm_gpu/src/sparse_ops/sparse_invert_permute.cu#L15

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    for (int32_t i = tid; i < segment_length; i += threads_per_segment) {
+        permuted_indices_[output_start + i] = indices_[input_start + i];
+    }
+}
+
+/**
+ * @brief Permute1DDataWithWeightsKernel operator implementation
+ */
+template <typename offsets_t, typename indices_t, typename weights_t>
+void Permute1DDataWithWeightsKernel<offsets_t, indices_t, weights_t>::operator()(
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp]: Why not to follow CUDA reference [1] and additionally parameterize with `has_weight`?

[1] https://github.com/pytorch/FBGEMM/blob/8744183e4bd7390a674a0d8736bf4d7ef2d0516e/fbgemm_gpu/src/sparse_ops/sparse_permute_1d.cu#L30

File: packages/fbgemm-xpu/src/fbgemm_xpu/ops.py [hunk 4]
```diff
@@ -5,12 +5,34 @@
...
 # This module provides user-friendly interfaces to the C++ operators
 
+from typing import Optional, Tuple
+
+import torch
+from torch import Tensor
+
 __all__ = [
-	"dense_embedding_codegen_lookup_function",
+    "invert_permute",
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/ops.py]: Here I am confused. Why do we expose these XPU specific APIs from fbgemm xpu? who is supposed to use them? My understanding was that we are just register low level entrypoints for XPU key so that standard fbgemm APIs start to work on XPU device. Shouldn't that be that user calls `fbgemm::invert_permute` for example?

File: packages/fbgemm-xpu/src/fbgemm_xpu/ops.py [hunk 5]
```diff
@@ -5,12 +5,34 @@
...
+    permuted_lengths_sum: Optional[int] = None,
+) -> Tuple[Tensor, Tensor, Optional[Tensor]]:
+    """Permutes sparse data in jagged/1D format according to permutation indices."""
+    return torch.ops.fbgemm.permute_1D_sparse_data.default(
+        permute, lengths, indices, weights, permuted_lengths_sum
+    )
+
+def asynchronous_complete_cumsum(t_in: Tensor) -> Tensor:
+    """Computes complete cumulative sum: output[0] = 0, output[i] = sum(t_in[0:i])."""
+    return torch.ops.fbgemm.asynchronous_complete_cumsum.default(t_in)
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/ops.py]: add empty line in end of file

File: packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
+"""
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py]: Should also have a copyright.

File: packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+Test suite for asynchronous_complete_cumsum operator
+
+Tests complete cumulative sum computation with leading zero.
+For 1D input [a, b, c], returns [0, a, a+b, a+b+c].
+For 2D input, cumsum along dimension 1 (columns).
+"""
+
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py]: Tests are eventually good, but our side tests won't guarantee that what we've implemented correctly integrates with the upstream FBGEMM. We must explore if upstream tests cover what we are implementing and use them if they are. If not, then we need to raise PRs with upstream to add test coverage. That's critical concern - we need to resolve that before proceeding.

File: packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+SEED = 42
+
+
+class TestAsynchronousCompleteCumsum(TestCase):
+    def setUp(self):
+        super().setUp()
+        torch.manual_seed(SEED)
+
+    def test_basic_int32(self):
+        if not torch.xpu.is_available():
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/test/test_asynchronous_complete_cumsum.py]: This check is inconsistent with checks in other test file: `@unittest.skipIf(not torch.xpu.is_available(), "XPU not available")`.

**Review by dvrogozh — CHANGES_REQUESTED** [review]
In addition to inline comments:
1. Repository has lint CI - please, address identified issues.
2. What we implement and merge must be tested under CI. And IMHO doing that by engineers who implemented the code is the best variant rather than offloading that to dedicated CI engineers. Thus, please, modify `.github/workflows/ci.yml` and add a test section for FBGEMM following the pattern outlined for torchcodec (see link below). The test job for FBGEMM will be much simpler. Target only `bmg` for now is fine. Note that fbgemm wheel is already getting build by CI.

https://github.com/intel/torchlib-xpu/blob/d4230aae4220649c6b879e43e2af166942abeebf/.github/workflows/ci.yml#L199

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 9]
```diff
@@ -118,5 +121,53 @@ index d358f4b8f..a462d45ed 100644
...
+ 
+ class CumSumTest(unittest.TestCase):
+     @given(
+@@ -40,7 +45,10 @@ class CumSumTest(unittest.TestCase):
+                 (torch.float32, np.float32),
+             ]
+         ),
+-        device=cpu_and_maybe_gpu(),
++        device=st.one_of(
++            cpu_and_maybe_gpu(),
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: Maybe we can just modify `cpu_and_maybe_gpu` to also return xpu if it is available?

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 10]
```diff
@@ -118,5 +121,53 @@ index d358f4b8f..a462d45ed 100644
...
++            st.sampled_from([torch.device("xpu")] if xpu_available else []),
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: That's strange construction, why `st.sampled_from` is needed at all? From the name this seems to be a random sampling primitive, but we have just single value in the list, so it does not make quite sense.

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 9 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: Changed in commit [0892f7f](https://github.com/intel/torchlib-xpu/pull/73/commits/0892f7f3c7c54865b4c0ff73506f619ce1f308d6)

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 10 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: My mistake, I misunderstood the functionality of  "st.sampled_from". This is corrected in commit [0892f7f](https://github.com/intel/torchlib-xpu/pull/73/commits/0892f7f3c7c54865b4c0ff73506f619ce1f308d6).

File: packages/fbgemm-xpu/src/fbgemm_xpu/sparse_async_cumsum.h [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+// HOST FUNCTION MAPPING:
+//   asynchronous_complete_cumsum_xpu (SYCL)
+//     → asynchronous_complete_cumsum_gpu (CUDA)
+//   asynchronous_exclusive_cumsum_xpu (SYCL)
+//     → asynchronous_exclusive_cumsum_gpu (CUDA)
+//   asynchronous_inclusive_cumsum_xpu (SYCL)
+//     → asynchronous_inclusive_cumsum_gpu (CUDA)
+//
+// DESCRIPTION:
+//   Computes cumulative sums on Intel XPU devices using PyTorch's native
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sparse_async_cumsum.h:26]: > using PyTorch's native cumsum operation

That's in distinction to https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/src/sparse_ops/sparse_async_cumsum.cu where operation is done via CUDA kernel, right? Can you extend a comment and emphasize that? Maybe better to place that under TODO in a comment. I guess this would depend on relative performance of the current implementation.

Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sparse_async_cumsum.h:26]: Yes, this version uses PyTorch instead of CUDA. One of the reasons for this is that CUDA version uses CCCL's "cub::DeviceScan::ExclusiveSum" function and not a FBGEMM kernel. [CCCL kernels](https://github.com/NVIDIA/cccl/blob/ff893069cb5db42c2552762466fd5c57c83da083/cub/cub/device/device_scan.cuh#L376) have a different architecture compared to FBGEMM kernels, and to adapt ExclusiveSum for FBGEMM would have taken more time than just implementing a simple cumulative sum with PyTorch for testing that the DLRM-v3 model works in XPU using TorchRec.
I'll add the TODO for future operator optimizations .

Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sparse_async_cumsum.h:26]: Added in commit [3d04e41](https://github.com/intel/torchlib-xpu/pull/73/commits/3d04e41076d37ee1c4aab24f3d4079891cdda916).
